src/test/java/org/folio/rest/api/RequestBatchAPITest.java

```diff
@@ -0,0 +1,192 @@
+package org.folio.rest.api;
+
+import static org.folio.rest.api.RequestsApiTest.requestStorageUrl;
+import static org.folio.rest.api.StorageTestSuite.TENANT_ID;
+import static org.folio.rest.api.StorageTestSuite.storageUrl;
+import static org.folio.rest.support.ResponseHandler.empty;
+import static org.folio.rest.support.ResponseHandler.json;
+import static org.folio.rest.support.builders.RequestRequestBuilder.OPEN_AWAITING_PICKUP;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+
+import java.net.URL;
+import java.util.Arrays;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.folio.rest.support.ApiTests;
+import org.folio.rest.support.JsonResponse;
+import org.folio.rest.support.Response;
+import org.folio.rest.support.builders.RequestRequestBuilder;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+public class RequestBatchAPITest extends ApiTests {
+
+  @Before
+  public void beforeEach() throws Exception {
+    StorageTestSuite.deleteAll(requestStorageUrl());
+  }
+
+  @After
+  public void checkIdsAfterEach() {
+    StorageTestSuite.checkForMismatchedIDs("request");
+  }
+
+  @Test
+  public void canUpdateRequestPositionsInBatch() throws Exception {
+    UUID itemId = UUID.randomUUID();
+
+    // 1st create some sample requests for an item
+    JsonObject firstRequest = createRequestForItemAtPosition(itemId, 1);
+    JsonObject secondRequest = createRequestForItemAtPosition(itemId, 2);
+
+    // 2nd: Swap positions for these requests
+    JsonObject firstRequestWithNoPosition = firstRequest.copy();
+    JsonObject secondRequestWithNoPosition = secondRequest.copy();
+
+    // 2.1 - release previous positions
+    secondRequestWithNoPosition.remove("position");
+    firstRequestWithNoPosition.remove("position");
+
+    // 2.2 - set new positions
+    firstRequest.put("position", 2);
+    secondRequest.put("position", 1);
```

Comment:
Why do the positions need removing? Does the `JsonObject.put` method not overwrite an existing property value?

What do the numbers in the comments refer to? 

Could the comments be replaced with extracted methods, if there are concerns about the individual steps not conveying the intent clearly enough?

Reply:
> Why do the positions need removing? Does the JsonObject.put method not overwrite an existing property value?

We have to release the position before assigning it to another request. This is because position constraint are being evaluated after each row update.

> What do the numbers in the comments refer to?

They refers to a flow step.

> Could the comments be replaced with extracted methods, if there are concerns about the individual steps not conveying the intent clearly enough?

For me, it increases code readability when we have such comments, so I can understand what this test do without diving into code.

Reply:
> > Why do the positions need removing? Does the JsonObject.put method not overwrite an existing property value?
> 
> We have to release the position before assigning it to another request. This is because position constraint are being evaluated after each row update.

Why do we have to release the position, I thought the intent of this batching was that clients no longer had to do that?

> > What do the numbers in the comments refer to?
> 
> They refers to a flow step.

What information does a flow step provide?

> > Could the comments be replaced with extracted methods, if there are concerns about the individual steps not conveying the intent clearly enough?
> 
> For me, it increases code readability when we have such comments, so I can understand what this test do without diving into code.

A big tradeoff for comments from my perspective is that they can easily get out of sync. I'd prefer extracting of well named methods that describe the flow at a higher level.

Reply:
> Why do we have to release the position, I thought the intent of this batching was that clients no longer had to do that?

No, we still have to release them, but it is more safe to failures during update: If there is a failure during position release/update all the changes will be reverted to initial state (as they were before reorder).

Unfortunately, there is no way to bypass the constraint, as by default unique constraints check are executed after each row update (not during transaction commit).

Reply:
Wait, what? I don't follow your reasoning here. Why can't we just remove the positions as part of the transaction? Why do we need to explicitly remove them as part of the input batch? Are we trying to avoid parsing the input JSON? I think it is pretty heavy handed to force the client to do this. We should be able to solve this all in the backend. The constraints are the storage module's problem. In fact, I would even say that it should be a validation error to pass in a position that is `null` or not set.

Maybe I am missing something here...

Reply:
@mreno-EBSCO, yep, make sense. But in this case the batch endpoint will serve only one particular case - request position update, that for me is a responsibility of the business module, and I'm not sure that it is fine to introduce such endpoint in storage module. 

> We should be able to solve this all in the backend. 

I don't think that this endpoint will be executed directly from UI, but as a part of an business flow.

What do you think?

Reply:
> But in this case the batch endpoint will serve only one particular case - request position update, that for me is a responsibility of the business module, and I'm not sure that it is fine to introduce such endpoint in storage module.

But isn't that exactly what this API is for, moving requests in the queue in an atomic way? I don't think it would be/should be used for anything else. Is that really what we are trying to accomplish here, a general purpose batch API to update requests? Is there a need for that?

I think we should be creating a specific POST endpoint for updating request queues of a specific item in an atomic way and leave it at that.

> I don't think that this endpoint will be executed directly from UI, but as a part of an business flow.

These are public endpoints. Any client with the correct permissions can call them and do whatever they are allowed to by the API, potentially updating every request in this case. This is one of the reasons why I think we need a very specific endpoint for request queue reordering.

I thought we were going to have a simple API, perhaps based on an itemId, where we pass in JSON that is just an array of objects containing the request ID and the new position, not the complete requests, and the storage module would do the reordering in a transaction. A general purpose batch API will require so much validation to ensure nothing is being corrupted. I just don't see the feasibility here.

I must be missing something...

Reply:
@mreno-EBSCO, I totally agree with you're statements above.

@marcjohnson-kint, are you ok with this approach? Introduce an endpoint in storage module that will supply the same reorderedQueue as business service and do update of positions only.

Reply:
Oh, I see. We are going to do a bunch of this work in `mod-circulation`. I still think this should not be a general purpose batch API.

Reply:
@bohdan-suprun I'm ok with a dedicated change positions endpoint, as long as it is in a separate interface, and I would like keep it experimental.

A concern I have is that it might appear similar to the higher level business logic endpoint, and that might be confusing. We want folks to use the business logic API, yet we cannot stop them using the storage API (I can refer folks to my comments elsewhere about some of the challenges with this)

Alas, as @mreno-EBSCO stated above, these are public APIs these kinds of aspects are inherent in the storage and business logic split.

Reply:
I don't want to change scope here, so maybe my comments are not appropriate.

That being said... Do we need an endpoint in the BL module for this? If so, couldn't it just be a pass through? I think anything we do in the BL module, we'd need to do in the storage module.

Reply:
I think we need do consider how much logic should be in storage, for example, should the storage API include validation for contiguous positions? Or that requests in certain statuses or types should not move from position 1?

Reply:
> ... should the storage API include validation for contiguous positions? Or that requests in certain statuses or types should not move from position 1?

Ah, good point.

I am always nervous about the faith we have in clients not to call storage APIs with inconsistent but valid data.

Reply:
I think it might be some issues with moving validation to storage module. Because in different cases we can allow or disallow same things.

For example for queue reorder we don't allow requests with started fulfillment to be moved from the 1st position. But this might be allowed when we chek-in an item and shift all requests in the queue. 

But I may be wrong.

Reply:
Yeah, I agree, there are challenges here given how much business logic is required to reorder the queue and the need to make this update atomic. My initial comments were a bit off, since I forgot the simpler API would be in mod-circulation.

However, I am concerned with how powerful the batch API is and that we might get in trouble with end users using it instead of mod-circulation thinking it is how they could bulk load request updates into the system. Then again, I'm not sure if that is something anyone would want to do.

Reply:
@mreno-EBSCO As I understand it, FOLIO as an architecture has decided to introduce batch APIs for both storage and business logic, so they are likely coming.

What I'm most concerned about is that we get a thorough understanding and agreement on the behaviour and design of that before we build too many of them. Which is why all of the existing ones are hopefully marked experimental and kept separate.

I think the topic of storage vs. business logic API use is an increasingly important and broad one, especially as folks start to build migrations tooling. I'm going to add it to the tech leads topic list.

@mreno-EBSCO @bohdan-suprun I think we need to decide whether to introduce a specific position only storage API or a very rough and experimental general batch API?

If we want to support update of the queue during fulfilment, then we likely need a more general API.

Reply:
On the subject of removing the positions first. I have two questions:

Why do we need to express this separate step in the tests, as I understand it, that is only relevant in the implementation?

Can we use [deferrable constraints](https://hashrocket.com/blog/posts/deferring-database-constraints) to avoid this in the implementation as well?

Reply:
Yep, deferrable constraints is a good option, but it seems RMB does not support it out of the box, so we will need to add it in a separate script  + care about migration (remove old constraint and replace it with the new one).

> Why do we need to express this separate step in the tests, as I understand it, that is only relevant in the implementation?

Well, currently the API assumes that client better know how to execute particular update (i mean order of updates, dealing with constraints, etc.) to get it successfully committed.

Reply:
@bohdan-suprun thanks for investigating the deferred constraints

Reply:
Hmm... I think we will be able to detect the fact that we're going to change positions in the batch, but this will require additional query to db (so we will compare provided positions vs positions in DB). And further, we can add a custom logic for this particular case.

@marcjohnson-kint, @mreno-EBSCO, what do you think?

Reply:
bq. The client have to take care about satisfying all the constraints (as in this case release all positions before assigning to pass item-position constraint). In a fact client should be aware about DB constraints, data consistency. etc.

If we are expecting the client to release the positions and then re-specify the new ones as two separate operations, what are we gaining by introducing this new storage API?

I think it is reasonable that the client satisfy the expectations that positions should be unique, I think we are leaking too much for a new batch API if we expect them to understand they need to do this in two operations.

I had thought we were trying to hide this from the client (in this case, mod-circulation).

> Hmm... I think we will be able to detect the fact that we're going to change positions in the batch, but this will require additional query to db (so we will compare provided positions vs positions in DB). And further, we can add a custom logic for this particular case.

I'm starting to understand how this could be complicated, because the batch API could be used for a mixture of requests for different items, or for only parts of a queue.

I'm beginning to wonder if we are ready to try to design a better storage API to support atomic updates with these constraint limitations?

Reply:
> If we are expecting the client to release the positions and then re-specify the new ones as two separate operations, what are we gaining by introducing this new storage API?

Well, we don't need two operations actually. We still need one operation, but batch package have to include requests where positions are removed and requests with the new positions set.

> I'm beginning to wonder if we are ready to try to design a better storage API to support atomic updates with these constraint limitations?

Maybe not, but from my point of view it more reliably then sending those updates concurrently in different transactions.

Reply:
bq. Well, we don't need two operations actually. We still need one operation, but batch package have to include requests where positions are removed and requests with the new positions set.

I think I need an example. Let's say we have a request queue for Harry Potter and the Cursed Child:
Request A, Position 1
Request B, Position 2
Request C, Position 3

Can the batch API be used, with a single operation, to perform the following reorder?
Request A, Position 1
Request C, Position 2
Request B, Position 3

Reply:
Yep, the batch API will look like follow:

Request A: with no position 
Request C: with no position 
Request B: with no position 
Request A, Position 1
Request C, Position 2
Request B, Position 3

As you can see, It includes 3 additional copies of the requests with positions removed. They must be executed before setting new positions.

Reply:
Hang on, we are expecting a client to provide two representations of the same record within the same batch API request?

I really don't think we want to expose that complexity to the clients for a new API

Reply:
Then the only option to implement API that will update positions only.

Reply:
If we accept this limitation in the current batch API, which I think is ok. Then I think we should do it in two operations.

The first to clear the positions, the second to update them. I know that isn't ideal, as it isn't atomic. I think it is simpler to understand than this kind of request.

@mreno-EBSCO  @mtraneis @jsavell what do you think?

Reply:
I think we might want to take a step back and decide if we really want to invest more in this right now, given the complications of the implementation

Reply:
> If we accept this limitation in the current batch API, which I think is ok. Then I think we should do it in two operations.

Executing it as two operations won't give us the main benefit of the batching. As @mreno-EBSCO mentioned in https://github.com/folio-org/mod-circulation/pull/366, it will be not possible to recover in case when removal executed successfully but position update has been failed. That's really the main reason to use this API (at least for the queue reorder feature).

Reply:
Then I think we should either abort this work and accept the current limitations, that failures can lead to a corrupted queue, or build a dedicated reordering API in storage with very limited behaviour.

I don’t think we are ready for a general purpose batch API, and submitting two representations of the same record and relying on it doing separate updates in order feels like it could be really brittle to me.

@mreno-EBSCO @skoczko what do you think?

I wonder if we’ve already invested more time than we really had in this

Reply:
So, what harm is done if we simply remove the `position` field of all the requests in the batch from what is stored in the DB prior to writing the batch? We do this when requests are expired. If we are assuming that the client is doing the right thing prior to calling the API then the existing positions should not matter. If they do, then the client missed something or the queue was modified after the client received the list, but before it sent the request and likely this will lead to a DB constraint violation and a transaction rollback.

Anyway, removing the positions as the first part of the transaction shouldn't matter even in a general purpose batch API. I don't think this is business logic, so much as requirements to write to the DB in this storage module. I don't think clients need to know this, nor should they, since there is no real way to convey this (maybe the API doc, but that isn't ideal). Even if positions are not changed, they could still be removed, since the request object being written will have the correct position already assigned, if not DB rollback, no harm...

Reply:
@marc-johnson Corrupting the queue leads to problems elsewhere and is confusing in the UI. We've already had to fix queues several times for Chalmers. Manually fixing queues will not scale. We cannot corrupt the queue. Chalmers gave an item to a patron without a loan because the patron was not next in the request queue, as they should have been, because of a corrupt queue. So, I do think we need to address this with some sort of transaction based reordering. Like I said above, I think if we just remove the `position` field for each request in the batch as the first part of the transaction, then I think we'll be OK. What do you think about that approach?

Reply:
@mreno-EBSCO i nearly raised the idea of removing the position internally to the batch API this morning.

I aborted as I started thinking of the various ways it could be possible for the API request to fail, and then got involved in other things.

I think we need to clearly document this behaviour. I agree it is much cleaner than submitting the records twice.

Let’s try it, I can’t remember why I shied away from it this morning.

Reply:
Hi @marcjohnson-kint,

Are we proceeding with the option when the batch API will detect whether positions has been changed or not and then add request representations with removed positions before updating? 

Another option that we have, if we agreed that supplying two representations of the same request is unacceptable, is following. 

Get rid of json-based unique index, but create two table columns instead (`itemId` and `position`) which will be populated by a trigger. So then we can create `deferrable` constraint and make things much easier. Is it make sense?

Reply:
@bohdan-suprun What do you think of the idea that the batch API removes the positions from all requests submitted, as a first step, and then does the regular updates, all within the implementation?

My understanding of that approach would be something like:

Let's say we have a request queue for Harry Potter and the Cursed Child:
Request A, Position 1
Request B, Position 2
Request C, Position 3

When the following requests are submitted to the batch API
Request A, Position 1
Request C, Position 2
Request B, Position 3

Internally the steps are as follows:

1. Start transaction
2. Remove the positions from all requests involved in the batch
3. Update the requests as usual
4. Commit transaction

This will in effect be an imitation of deferred constraint.

I think this is much cleaner and better reflects the general purpose nature of the API.

I had reservations about suggesting that yesterday, as I started to get concerned about what happens if the API is used for partial queue updates or requests for multiple items, but can't seem to think of specific issues, as every request whose position changes should be involved in the batch submission.

I think we need to make sure we document the reasons behind this and the limitations.

Reply:
I think if that approach doesn't work well, then I think we should go with the approach you suggested and swap the expression based constraint for triggers and a regular, deferrable constraint, even though I'm not a fan of triggers.

Reply:
> What do you think of the idea that the batch API removes the positions from all requests submitted, as a first step, and then does the regular updates, all within the implementation?

I can not find any issues with such approach. I'm fine with this.

Reply:
I've committed the changes that you've suggested. I have couple of questions:
1. Jenkins says following in build logs:
```
WARNING: lint-raml-cop: Missing from configuration: request-storage-batch.raml
WARNING: lint-raml-cop: Configuration needs to be updated (FOLIO-903).
```
Should I create a PR to the configuration repo?

2. 
> I think we need to make sure we document the reasons behind this and the limitations.

Do we have a separate documentation somewhere on wiki? Or you mean Java docs?

Reply:
> Should I create a PR to the configuration repo?

I think @dcrossleyau is best placed to answer that question. It can be tested locally, however the global configuration should only be changed after this is merged. As once it is added, the documentation build will expect that RAML file to be present.

> Do we have a separate documentation somewhere on wiki? Or you mean Java docs?

I am referring to the Java docs or the readme / documentation within the repository. I don't think implementation specific information should be kept on the wiki.

Reply:
Thanks Marc. Yes the warnings about extra configuration can be ignored.

All RAML files are still automatically discovered and processed.

Also after merge to master, all API docs (including new ones) will still be generated and published. But as explained in the configuration documentation, they will just not be linked from the [API docs](https://dev.folio.org/reference/api/) table.

We could adjust the configuration for every new RAML file added via a branch. However then there would be a warning message in the opposite direction for the other branches, about configuration not matching the discovered set of files. Note that i manually follow up each day, to add new configuration for all master branches.

By the way, all of these warning and info messages would all be gone from the PR screen, if people attended to the missing schema descriptions.